katz/src/main/kotlin/katz/Reader.kt

```diff
@@ -0,0 +1,38 @@
+package kats
+
+/**
+ * Basic implementation of the Reader monad. Provides an "implicit" context (configuration) for
+ * function execution. Intended to provide Dependency Injection.
+ */
+class Reader<C : Any, out A : Any>(val run: (C) -> A) {
+
+    inline fun <B : Any> map(crossinline fa: (A) -> B): Reader<C, B> = Reader { c -> fa(run(c)) }
+
+    inline fun <B : Any> flatMap(crossinline fa: (A) -> Reader<C, B>): Reader<C, B> = Reader { c ->
```

Comment:
This is the way if you want set `C` as contravariant type:

```
inline fun <C : Any, A : Any, B : Any> Reader<C, A>.flatMap(crossinline fa: (A) -> Reader<C, B>): Reader<C, B> = 
        Reader { c -> fa(run(c)).run(c) }
```

Remove `flatMap` and implement it as an extension function.